collector/mdadm_linux.go

```diff
@@ -18,199 +18,141 @@ package collector
-		return []mdStatus{}, err
-	}
-
-	lines := strings.Split(string(content), "\n")
-	// Each md has at least the deviceline, statusline and one empty line afterwards
-	// so we will have probably something of the order len(lines)/3 devices
-	// so we use that for preallocation.
-	mdStates := make([]mdStatus, 0, len(lines)/3)
-	for i, line := range lines {
-		if line == "" {
-			continue
-		}
-		if line[0] == ' ' || line[0] == '\t' {
-			// Lines starting with white space are not the beginning of a md-section.
-			continue
-		}
-		if strings.HasPrefix(line, "Personalities") || strings.HasPrefix(line, "unused") {
-			// These lines contain general information.
-			continue
-		}
-
-		mainLine := strings.Split(line, " ")
-		if len(mainLine) < 4 {
-			return mdStates, fmt.Errorf("error parsing mdline: %s", line)
-		}
-		md := mdStatus{
-			name:   mainLine[0],
-			active: mainLine[2] == "active",
-		}
-
-		if len(lines) <= i+3 {
-			return mdStates, fmt.Errorf("error parsing mdstat: entry for %s has fewer lines than expected", md.name)
-		}
-
-		personality := ""
-		for _, possiblePersonality := range mainLine[3:] {
-			if raidPersonalityRE.MatchString(possiblePersonality) {
-				personality = possiblePersonality
-				break
-			}
-		}
-		switch {
-		case personality == "raid0":
-			md.disksActive = int64(len(mainLine) - 4) // Get the number of devices from the main line.
-			md.disksTotal = md.disksActive            // Raid0 active and total is always the same if active.
-			md.blocksTotal, err = evalRaid0line(lines[i+1])
-		case raidPersonalityRE.MatchString(personality):
-			md.disksActive, md.disksTotal, md.blocksTotal, err = evalStatusline(lines[i+1])
-		default:
-			log.Infof("Personality unknown: %s\n", mainLine)
-			md.blocksTotal, err = evalUnknownPersonalitylineRE(lines[i+1])
-		}
-		if err != nil {
-			return mdStates, fmt.Errorf("error parsing mdstat: %s", err)
-		}
-
-		syncLine := lines[i+2]
-		if strings.Contains(syncLine, "bitmap") {
-			syncLine = lines[i+3]
-		}
-
-		// If device is syncing at the moment, get the number of currently synced bytes,
-		// otherwise that number equals the size of the device.
-		if strings.Contains(syncLine, "recovery") ||
-			strings.Contains(syncLine, "resync") &&
-				!strings.Contains(syncLine, "\tresync=") {
-			md.blocksSynced, err = evalBuildline(syncLine)
-			if err != nil {
-				return mdStates, fmt.Errorf("error parsing mdstat: %s", err)
-			}
+	file.Close()
+
+	md.disksActive = float64(array.active_disks)
+	md.disksFailed = float64(array.failed_disks)
+	md.disksMissing = float64(max(0, array.raid_disks-array.nr_disks))
+	md.disksSpare = float64(array.spare_disks)
+	md.bytesTotal = float64(devsize)
+	//TODO make md.active bool or document state (md.active)
+	md.active = float64(array.state)
+
+	sys_sync_completed := sysFilePath("block/" + md.name + "/md/sync_completed")
+	log.Debugf("sys_sync_completed %s", sys_sync_completed)
+	content, err := ioutil.ReadFile(sys_sync_completed)
+	if err == nil {
```

Comment:
I say check for the error here instead of checking if it's nil.  By inverting the logic and adding an early return, you can save a level of indentation and have safer code that doesn't fail silently.